### PR TITLE
Add GrpcOrGrpcWebClientChannel (#457)

### PR DESCRIPTION
**Pull request by morgwai**

### CODE REVIEW

File: test/client_tests/grpc_or_grpcweb_channel_grpc_test.dart [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      host: host,
+      grpcPort: grpcPort,
+      grpcWebPort: grpcWebPort,
+      secure: false,
+    );
+    expect(channel is ClientChannel, isTrue);
+    final grpcChannel = channel as ClientChannel;
+    expect(grpcChannel.host, equals(host));
+    expect(grpcChannel.port, equals(grpcPort));
+    expect(grpcChannel.options.credentials.isSecure, isFalse);
```
Comment (morgwai) [inline test/client_tests/grpc_or_grpcweb_channel_grpc_test.dart:37]: I was considering to check also if all other options have default values, but not sure if it's a good idea: please advice.

**Review by morgwai — COMMENTED** [review]
3 questions to the reviewers

**Comment by morgwai** [comment]
size XXL is due to generated/copied proto files (~700 lines). The real code including tests and example is ~350 lines.

**Comment by morgwai** [comment]
looking at the number of warnings generated by the example, I'm wondering if it would be better to move it to a separate repo: please advise.

File: test/client_tests/grpc_or_grpcweb_channel_grpc_test.dart [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+void main() {
+  test('Channel on non-web uses gRPC ClientChannel with correct params', () {
+    final channel = GrpcOrGrpcWebClientChannel(
+      host: host,
+      grpcPort: grpcPort,
+      grpcWebPort: grpcWebPort,
+      secure: false,
+    );
+    expect(channel is ClientChannel, isTrue);
+    final grpcChannel = channel as ClientChannel;
```
[the remark this answers is not in the document]
Comment (mraleph) [reply test/client_tests/grpc_or_grpcweb_channel_grpc_test.dart:34]: This test is not going to run in the browser (given the annotation `@TestOn`) which I think means that you can just drop the cast.

File: test/client_tests/grpc_or_grpcweb_channel_grpc_test.dart (hunk 1 above)
Comment (mraleph) [reply test/client_tests/grpc_or_grpcweb_channel_grpc_test.dart:37]: I think it is okay to just test a subset.

File: example/grpc_or_grpcweb/lib/src/generated/echo.pbjson.dart [hunk 3]
```diff
@@ -0,0 +1,80 @@
+///
+//  Generated code. Do not modify.
+//  source: echo.proto
+//
+// @dart = 2.12
+// ignore_for_file: annotate_overrides,camel_case_types,unnecessary_const,non_constant_identifier_names,library_prefixes,unused_import,unused_shown_name,return_of_invalid_type,unnecessary_this,prefer_final_fields,deprecated_member_use_from_same_package
+
+import 'dart:core' as $core;
+import 'dart:convert' as $convert;
```
[the remark this answers is not in the document]
Comment (mraleph) [reply example/grpc_or_grpcweb/lib/src/generated/echo.pbjson.dart:9]: I think you should just expand `ignore_for_file` line above to suppress this lint. 

/fyi @pq `protoc` generated code is triggering the lint which you have recently strengthened up :(

**Comment by mraleph** [comment]
@morgwai I think you should be able to make an example which does not depend on Flutter or just drop the example altogether for now.

File: test/client_tests/grpc_or_grpcweb_channel_grpc_test.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (morgwai) [reply test/client_tests/grpc_or_grpcweb_channel_grpc_test.dart]: !
I'm an idiot ;-)  I got biased by the fact that I wrote the web test first, where cast was necessary ;-]
Thanks :)

File: example/grpc_or_grpcweb/lib/src/generated/echo.pbjson.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (morgwai) [reply example/grpc_or_grpcweb/lib/src/generated/echo.pbjson.dart]: I've removed the example for now.

**Comment by morgwai** [comment]
@mraleph thanks for the comments :-)  Please have another look in a spare moment.

**Review by mraleph — APPROVED** [review]

**Comment by mraleph** [comment]
One last request: please add an item to CHANGELOG describing the change. I think it is a great change to include into the changelog.

**Comment by morgwai** [comment]
@mraleph  Passed :) I don't see a 'merge' button or anything: I think you need to merge it.

**Comment by morgwai** [comment]
> One last request: please add an item to CHANGELOG describing the change. I think it is a great change to include into the changelog.

ah yes: 3 minutes please :)

**Comment by morgwai** [comment]
@mraleph should I add under `3.0.1-dev` version, or some other?

**Comment by mraleph** [comment]
Yep under the 3.0.1-dev.

**Comment by morgwai** [comment]
@mraleph pushed, please start the workflow.

**Review by mraleph — APPROVED** [review]

**Comment by morgwai** [comment]
@mraleph  ready to merge :)  Many thanks for the review!

File: example/grpc_or_grpcweb/lib/src/generated/echo.pbjson.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (pq) [reply example/grpc_or_grpcweb/lib/src/generated/echo.pbjson.dart]: Sorry, yeah.  The fix up to `directives_ordering` was a lot more disruptive to the ecosystem than we expected (ping https://github.com/dart-lang/linter/issues/2609). 😬 

As for the remedy, ignoring the lint in generated sources is not terrible in the short term but longer term, for sure, it should be fixed in the code generator.

https://github.com/dart-lang/protobuf/issues/511

**Comment by ericbets** [comment]
Can anyone comment on what happened to this? It says merged but I can't find it anywhere.

**Comment by mraleph** [comment]
@ericbets it's right here: https://github.com/grpc/grpc-dart/blob/master/lib/grpc_or_grpcweb.dart, you need to be using 3.0.1 or later to get it.

**Pull request merged**